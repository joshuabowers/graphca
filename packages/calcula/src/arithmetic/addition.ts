
// const MpM = visit(Multiplication, Multiplication)
// const MpB = visit(Multiplication, Base)
// const BpM = visit(Base, Multiplication)

// export const add = binary(Addition)(
// )(
//   MpM(rightChild, rightChild)((l, r) => multiply(add(l.left, r.left), l.right)),
//   MpM(leftChild, rightChild)((l, r) => multiply(add(l.right, r.left), l.left)),
//   MpM(rightChild, leftChild)((l, r) => multiply(add(l.left, r.right), l.right)),
//   MpM(leftChild, leftChild)((l, r) => multiply(add(l.right, r.right), l.left)),
//   MpM(rightChild, negated(rightChild))((l, r) => (is(Binary)(r.right) && multiply(subtract(l.left, r.right.left), l.right)) || real(NaN)),
//   MpM(rightChild, negated(leftChild))((l, r) => (is(Binary)(r.right) && multiply(subtract(l.left, r.right.right), l.right)) || real(NaN)),
//   MpM(leftChild, negated(rightChild))((l, r) => (is(Binary)(r.right) && multiply(subtract(l.right, r.right.left), l.left)) || real(NaN)),
//   MpM(leftChild, negated(leftChild))((l, r) => (is(Binary)(r.right) && multiply(subtract(l.right, r.right.right), l.left)) || real(NaN)),
//   MpM(negated(rightChild), leftChild)((l, r) => (is(Binary)(l.right) && multiply(subtract(r.right, l.right.left), r.left)) || real(NaN)),
//   MpM(negated(rightChild), rightChild)((l, r) => (is(Binary)(l.right) && multiply(subtract(r.left, l.right.left), r.right)) || real(NaN)),
//   MpM(negated(leftChild), leftChild)((l, r) => (is(Binary)(l.right) && multiply(subtract(r.right, l.right.right), r.left)) || real(NaN)),
//   MpM(negated(leftChild), rightChild)((l, r) => (is(Binary)(l.right) && multiply(subtract(r.left, l.right.right), r.right)) || real(NaN)),

//   MpB(leftChild, identity)((l, r) => multiply(add(real(1), l.right), r)),
//   MpB(rightChild, identity)((l, r) => multiply(add(real(1), l.left), r)),
//   BpM(identity, leftChild)((l, r) => multiply(add(real(1), r.right), l)),
//   BpM(identity, rightChild)((l, r) => multiply(add(real(1), r.left), l))
// )

// export const subtract = binaryFrom(add, (l, r) => [l, negate(r)])

import { Writer, unit } from '../monads/writer'
import { _ } from '@arrows/multimethod'
import { 
  TreeNode, Genera, Species, Notation, any, notAny 
} from "../utility/tree"
import { 
  Real, real, complex, boolean, isPrimitive, PrimitiveNode 
} from "../primitives"
import { 
  Binary, binary, binaryFrom, when, binaryInfixRule 
} from "../closures/binary"
import { deepEquals, isValue } from "../utility/deepEquals"
import { multiply, double, negate, Multiplication, isMultiplication } from './multiplication'
import { rule, identityRule } from '../utility/rule'

export type Addition = Binary<Species.add, Genera.arithmetic>
type AdditionWithPrimitive = Addition & {
  readonly right: Writer<PrimitiveNode>
}

export const addRule = binaryInfixRule('+')

export const [add, isAddition, $add] = binary<Addition>(
  '+', Notation.infix, Species.add, Genera.arithmetic
)(
  (l, r) => real(l.value + r.value),
  (l, r) => complex([l.a + r.a, l.b + r.b]), 
  (l, r) => boolean((l.value || r.value) && !(l.value && r.value)), 
)(
  when([
    any(Species.real, Species.complex, Species.boolean), 
    notAny(Species.real, Species.complex, Species.boolean)
  ], (l, r) => [add(unit(r), unit(l)), addRule(r, l), 'reorder operands']),
  when<TreeNode, Real>([_, isValue(real(0))], (l, _r) => [l, identityRule(l), 'additive identity']),
  when<AdditionWithPrimitive, PrimitiveNode>(
    (l, r) => isAddition(l) 
      && isPrimitive(l.value.right) 
      && isPrimitive(r), 
    (l, r) => [
      add(l.left, add(l.right, unit(r))), 
      rule`${l.left} + (${l.right} + ${r})`,
      'additive associativity'
    ]),
  when(deepEquals, (l, _r) => [
    double(unit(l)), rule`2 * ${l}`, 'equivalence: replaced with double'
  ]),
  when<Addition, TreeNode>(
    (l, r) => isAddition(l) && deepEquals(l.value.left, r),
    (l, _r) => [add(double(l.left), l.right), rule`2 * ${l.left} + ${l.right}`, 'combined like terms']
  ),
  when<Addition, TreeNode>(
    (l, r) => isAddition(l) && deepEquals(l.value.right, r),
    (l, _r) => [add(double(l.right), l.left), rule`2 * ${l.right} + ${l.left}`, 'combined like terms']
  ),
  when<TreeNode, Addition>(
    (l, r) => isAddition(r) && deepEquals(l, r.value.left),
    (_l, r) => [add(double(r.left), r.right), rule`2 * ${r.left} + ${r.right}`, 'combined like terms']
  ),
  when<TreeNode, Addition>(
    (l, r) => isAddition(r) && deepEquals(l, r.value.right),
    (_l, r) => [add(double(r.right), r.left), rule`2 * ${r.right} + ${r.left}`, 'combined like terms']
  ),
  when<Multiplication, Multiplication>(
    (l, r) => isMultiplication(l) && isMultiplication(r)
      && isPrimitive(l.value.left) && isPrimitive(r.value.left)
      && deepEquals(l.value.right, r.value.right),
    (l, r) => [
      multiply(add(l.left, r.left), l.right), 
      rule`(${l.left} + ${r.left}) * ${l.right}`,
      'combined like terms'
    ]
  ),
  when<TreeNode, Multiplication>(
    (l, r) => isMultiplication(r) && isPrimitive(r.value.left) 
      && deepEquals(l, r.value.right),
    (l, r) => [multiply(add(real(1), r.left), unit(l)), rule`(${real(1)} + ${r.left}) * ${l}`, 'combined like terms']
  ),
  when<Multiplication, TreeNode>(
    (l, r) => isMultiplication(l) && isPrimitive(l.value.left) 
      && deepEquals(l.value.right, r),
    (l, r) => [multiply(add(real(1), l.left), unit(r)), rule`(${real(1)} + ${l.left}) * ${r}`, 'combined like terms']
  )
)

export const subtract = binaryFrom(add)(undefined, r => negate(r))
