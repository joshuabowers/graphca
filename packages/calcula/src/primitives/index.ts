import { isPrimitive, PrimitiveNode } from '../closures/primitive'
import { Real, real, isReal, EulerMascheroni } from './real'
import { 
  Complex, complex, isComplex, ComplexInfinity, isComplexInfinity 
} from './complex'
import { Boolean, boolean, isBoolean } from './boolean'
import { Nil, nil, isNil } from './nil'
import { NaN, nan, isNaN } from './nan'

export type Primitive = Real | Complex | Boolean | Nil | NaN
export { Real, Complex, Boolean, Nil, NaN }

export type PrimitiveFns = 
| typeof real | typeof complex | typeof boolean

export { real, complex, boolean, nil, nan }
export { isReal, isComplex, isBoolean, isNil, isNaN, isComplexInfinity }
export { isPrimitive, PrimitiveNode }
export { EulerMascheroni, ComplexInfinity }
