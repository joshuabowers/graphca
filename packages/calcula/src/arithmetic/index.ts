import { add, subtract, Addition, isAddition } from "./addition";
import { 
  multiply, divide, negate, double, Multiplication, isMultiplication 
} from "./multiplication";
import { 
  raise, reciprocal, square, sqrt, Exponentiation, isExponentiation 
} from "./exponentiation";

export { 
  add, subtract, 
  multiply, divide, negate, double,
  raise, reciprocal, square, sqrt
}
export {
  Addition, Multiplication, Exponentiation
}
export {
  isAddition, isMultiplication, isExponentiation
}
