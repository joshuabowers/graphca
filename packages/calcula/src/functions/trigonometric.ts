import { unit } from "../monads/writer"
import { Genera, Species, Notation, isGenus } from "../utility/tree"
import { real, complex, boolean } from "../primitives"
import { UnaryNode, unary } from "../closures/unary"
import { reciprocal } from "../arithmetic"
import { rule } from "../utility/rule"

export type TrigonometricNode = UnaryNode & {
  readonly genus: Genera.trigonometric
}

export const isTrigonometric = isGenus<TrigonometricNode>(Genera.trigonometric)

type Trigonometric<S extends Species> = TrigonometricNode & {
  readonly species: S
}

export type Sine = Trigonometric<Species.sin>
export type Cosine = Trigonometric<Species.cos>
export type Tangent = Trigonometric<Species.tan>
export type Cosecant = Trigonometric<Species.csc>
export type Secant = Trigonometric<Species.sec>
export type Cotangent = Trigonometric<Species.cot>

export const [sin, isSine, $sin] = unary<Sine>(
  'sin', Notation.prefix, Species.sin, Genera.trigonometric
)(
  r => real(Math.sin(r.value)),
  c => complex([
    Math.sin(c.a) * Math.cosh(c.b),
    Math.cos(c.a) * Math.sinh(c.b)
  ]),
  b => boolean(real(Math.sin(b.value ? 1 : 0))),
)()

export const [cos, isCosine, $cos] = unary<Cosine>(
  'cos', Notation.prefix, Species.cos, Genera.trigonometric
)(
  r => real(Math.cos(r.value)), 
  c => complex([
    Math.cos(c.a) * Math.cosh(c.b),
    -Math.sin(c.a) * Math.sinh(c.b)
  ]),
  b => boolean(real(Math.cos(b.value ? 1 : 0))),
)()

export const [tan, isTangent, $tan] = unary<Tangent>(
  'tan', Notation.prefix, Species.tan, Genera.trigonometric
)(
  r => real(Math.tan(r.value)),
  c => {
    const divisor = Math.cos(2 * c.a) + Math.cosh(2 * c.b)
    return complex([
      Math.sin(2 * c.a) / divisor,
      Math.sinh(2 * c.b) / divisor    
    ])
  },
  b => boolean(real(Math.tan(b.value ? 1 : 0))),
)()

export const [sec, isSecant, $sec] = unary<Secant>(
  'sec', Notation.prefix, Species.sec, Genera.trigonometric,
)(
  r => reciprocal(cos(real(r.value))),
  c => reciprocal(cos(complex([c.a, c.b]))),
  b => reciprocal(cos(boolean(b.value)))
)()

export const [csc, isCosecant, $csc] = unary<Cosecant>(
  'csc', Notation.prefix, Species.csc, Genera.trigonometric,
)(
  r => reciprocal(sin(real(r.value))),
  c => reciprocal(sin(complex([c.a, c.b]))),
  b => reciprocal(sin(boolean(b.value)))
)()

export const [cot, isCotangent, $cot] = unary<Cotangent>(
  'cot', Notation.prefix, Species.cot, Genera.trigonometric,
)(
  r => reciprocal(tan(real(r.value))),
  c => reciprocal(tan(complex([c.a, c.b]))),
  b => reciprocal(tan(boolean(b.value)))
)()
