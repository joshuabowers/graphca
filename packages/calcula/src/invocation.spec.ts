import { expectToEqualWithSnapshot } from './utility/expectations'
import { real, complex, boolean } from './primitives'
import { variable, scope } from './variable'
import { add } from './arithmetic/addition'
import { invoke } from './invocation'
import { multiply } from './arithmetic'
import { cos } from './functions/trigonometric'

describe('invoke', () => {
  describe('with no passed scope', () => {
    it('applies its arguments to unbound variables', () => {
      expectToEqualWithSnapshot(
        invoke()(add(variable('x'), variable('y')))(real(5), real(10)),
        real(15)
      )
    })

    it('applies its arguments in appearance order', () => {
      expectToEqualWithSnapshot(
        invoke()(add(variable('x'), variable('y')))(real(5)),
        add(variable('y'), real(5))
      )
    })
  })

  describe('with a passed scope but no arguments', () => {
    it('evaluates any unbound variables against its scope', () => {
      const s = scope([['x', variable('x', real(5))]])

      expectToEqualWithSnapshot(
        invoke(s)(add(variable('x'), variable('y')))(),
        add(variable('y'), real(5))
      )
    })
  })

  describe('with a passed scope and arguments', () => {
    it('evaluates any unbound variables against argument overridden scope', () => {
      const s = scope([['x', variable('x', real(5))]])

      expectToEqualWithSnapshot(
        invoke(s)(add(variable('x'), variable('y')))(real(10)),
        add(variable('y'), real(10))
      )
    })
  })

  describe('when given an expression with multiple instances of a variable', () => {
    it('substitutes a newly bound value for each variable instance', () => {
      expectToEqualWithSnapshot(
        invoke()(multiply(variable('x'), add(variable('x'), real(5))))(real(4)),
        real(36)
      )
    })
  })

  describe('of specific node types', () => {
    it('returns reals directly', () => {
      expectToEqualWithSnapshot(
        invoke()(real(5))(),
        real(5)
      )
    })

    it('returns complex numbers directly', () => {
      expectToEqualWithSnapshot(
        invoke()(complex([1, 5]))(),
        complex([1, 5])
      )
    })

    it('returns booleans directly', () => {
      expectToEqualWithSnapshot(
        invoke()(boolean(true))(),
        boolean(true)
      )
    })

    it('returns unbound variables directly', () => {
      expectToEqualWithSnapshot(
        invoke()(variable('x'))(),
        variable('x')
      )
    })

    it('returns the value of an argument-bound variable', () => {
      expectToEqualWithSnapshot(
        invoke()(variable('x'))(real(5)),
        real(5)
      )
    })

    it('returns the value of a scope-bound variable', () => {
      const s = scope([['x', variable('x', real(5))]])
      expectToEqualWithSnapshot(
        invoke(s)(variable('x'))(),
        real(5)
      )
    })

    it('substitutes a variable with an unbound variable argument', () => {
      expectToEqualWithSnapshot(
        invoke()(variable('x'))(variable('y')),
        variable('y')
      )
    })

    it('evaluates left child variables of binaries', () => {
      expectToEqualWithSnapshot(
        invoke()(add(variable('x'), real(5)))(real(10)),
        real(15)
      )
    })

    it('evaluates right child variables of binaries', () => {
      expectToEqualWithSnapshot(
        invoke()(multiply(real(5), variable('x')))(real(10)),
        real(50)
      )
    })

    it('evaluates child variables of unary nodes', () => {
      expectToEqualWithSnapshot(
        invoke()(cos(variable('x')))(real(Math.PI)),
        real(-1)
      )
    })
  })
})
