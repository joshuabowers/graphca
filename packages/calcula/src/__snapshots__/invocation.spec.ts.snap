// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`invoke of specific node types evaluates child variables of unary nodes 1`] = `
[
  [
    "(cos(x))(ùúã)",
    "identified invocation",
  ],
  [
    "([cos(x)])(ùúã)",
    "processing expression",
  ],
  [
    "cos(x)",
    "identified cosine",
  ],
  [
    "cos([x])",
    "processing argument",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "cos({x})",
    "processed argument",
  ],
  [
    "cos(x)",
    "created cosine",
  ],
  [
    "({cos(x)})(ùúã)",
    "processed expression",
  ],
  [
    "(cos(x))([ùúã])",
    "processing 1st",
  ],
  [
    "ùúã",
    "created real",
  ],
  [
    "(cos(x))({ùúã})",
    "processed 1st",
  ],
  [
    "(cos(x))(ùúã)",
    "cosine invocation",
  ],
  [
    "{x:=ùúã}",
    "established scope",
  ],
  [
    "cos(x)",
    "looking for variables to substitute in expression",
  ],
  [
    "cos(x)",
    "identified cosine",
  ],
  [
    "cos([x])",
    "processing argument",
  ],
  [
    "x",
    "substituting variable x",
  ],
  [
    "ùúã",
    "created real",
  ],
  [
    "cos({ùúã})",
    "processed argument",
  ],
  [
    "cos(ùúã)",
    "real cosine",
  ],
  [
    "-1",
    "created real",
  ],
]
`;

exports[`invoke of specific node types evaluates left child variables of binaries 1`] = `
[
  [
    "((x+5))(10)",
    "identified invocation",
  ],
  [
    "([(x+5)])(10)",
    "processing expression",
  ],
  [
    "(x+5)",
    "identified addition",
  ],
  [
    "([x]+5)",
    "processing left operand",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "({x}+[5])",
    "processed left operand; processing right operand",
  ],
  [
    "5",
    "created real",
  ],
  [
    "(x+{5})",
    "processed right operand",
  ],
  [
    "(x+5)",
    "created addition",
  ],
  [
    "({(x+5)})(10)",
    "processed expression",
  ],
  [
    "((x+5))([10])",
    "processing 1st",
  ],
  [
    "10",
    "created real",
  ],
  [
    "((x+5))({10})",
    "processed 1st",
  ],
  [
    "((x+5))(10)",
    "addition invocation",
  ],
  [
    "{x:=10}",
    "established scope",
  ],
  [
    "(x+5)",
    "looking for variables to substitute in expression",
  ],
  [
    "(x+5)",
    "identified addition",
  ],
  [
    "([x]+5)",
    "processing left operand",
  ],
  [
    "x",
    "substituting variable x",
  ],
  [
    "10",
    "created real",
  ],
  [
    "({10}+[5])",
    "processed left operand; processing right operand",
  ],
  [
    "5",
    "invoked Real",
  ],
  [
    "5",
    "created real",
  ],
  [
    "(10+{5})",
    "processed right operand",
  ],
  [
    "(10+5)",
    "real addition",
  ],
  [
    "15",
    "created real",
  ],
]
`;

exports[`invoke of specific node types evaluates right child variables of binaries 1`] = `
[
  [
    "((5*x))(10)",
    "identified invocation",
  ],
  [
    "([(5*x)])(10)",
    "processing expression",
  ],
  [
    "(5*x)",
    "identified multiplication",
  ],
  [
    "([5]*x)",
    "processing left operand",
  ],
  [
    "5",
    "created real",
  ],
  [
    "({5}*[x])",
    "processed left operand; processing right operand",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "(5*{x})",
    "processed right operand",
  ],
  [
    "(5*x)",
    "created multiplication",
  ],
  [
    "({(5*x)})(10)",
    "processed expression",
  ],
  [
    "((5*x))([10])",
    "processing 1st",
  ],
  [
    "10",
    "created real",
  ],
  [
    "((5*x))({10})",
    "processed 1st",
  ],
  [
    "((5*x))(10)",
    "multiplication invocation",
  ],
  [
    "{x:=10}",
    "established scope",
  ],
  [
    "(5*x)",
    "looking for variables to substitute in expression",
  ],
  [
    "(5*x)",
    "identified multiplication",
  ],
  [
    "([5]*x)",
    "processing left operand",
  ],
  [
    "5",
    "invoked Real",
  ],
  [
    "5",
    "created real",
  ],
  [
    "({5}*[x])",
    "processed left operand; processing right operand",
  ],
  [
    "x",
    "substituting variable x",
  ],
  [
    "10",
    "created real",
  ],
  [
    "(5*{10})",
    "processed right operand",
  ],
  [
    "(5*10)",
    "real multiplication",
  ],
  [
    "50",
    "created real",
  ],
]
`;

exports[`invoke of specific node types returns booleans directly 1`] = `
[
  [
    "(true)()",
    "identified invocation",
  ],
  [
    "([true])()",
    "processing expression",
  ],
  [
    "true",
    "created boolean",
  ],
  [
    "({true})()",
    "processed expression",
  ],
  [
    "(true)()",
    "boolean invocation",
  ],
  [
    "{}",
    "established scope",
  ],
  [
    "true",
    "invoked Boolean",
  ],
  [
    "true",
    "created boolean",
  ],
]
`;

exports[`invoke of specific node types returns complex numbers directly 1`] = `
[
  [
    "(1+5ùíä)()",
    "identified invocation",
  ],
  [
    "([1+5ùíä])()",
    "processing expression",
  ],
  [
    "1+5ùíä",
    "created complex",
  ],
  [
    "({1+5ùíä})()",
    "processed expression",
  ],
  [
    "(1+5ùíä)()",
    "complex invocation",
  ],
  [
    "{}",
    "established scope",
  ],
  [
    "1+5ùíä",
    "invoked Complex",
  ],
  [
    "1+5ùíä",
    "created complex",
  ],
]
`;

exports[`invoke of specific node types returns reals directly 1`] = `
[
  [
    "(5)()",
    "identified invocation",
  ],
  [
    "([5])()",
    "processing expression",
  ],
  [
    "5",
    "created real",
  ],
  [
    "({5})()",
    "processed expression",
  ],
  [
    "(5)()",
    "real invocation",
  ],
  [
    "{}",
    "established scope",
  ],
  [
    "5",
    "invoked Real",
  ],
  [
    "5",
    "created real",
  ],
]
`;

exports[`invoke of specific node types returns the value of a scope-bound variable 1`] = `
[
  [
    "(x)()",
    "identified invocation",
  ],
  [
    "([x])()",
    "processing expression",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "({x})()",
    "processed expression",
  ],
  [
    "(x)()",
    "variable invocation",
  ],
  [
    "{x:=5}",
    "established scope",
  ],
  [
    "x",
    "substituting variable x",
  ],
  [
    "5",
    "created real",
  ],
]
`;

exports[`invoke of specific node types returns the value of an argument-bound variable 1`] = `
[
  [
    "(x)(5)",
    "identified invocation",
  ],
  [
    "([x])(5)",
    "processing expression",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "({x})(5)",
    "processed expression",
  ],
  [
    "(x)([5])",
    "processing 1st",
  ],
  [
    "5",
    "created real",
  ],
  [
    "(x)({5})",
    "processed 1st",
  ],
  [
    "(x)(5)",
    "variable invocation",
  ],
  [
    "{x:=5}",
    "established scope",
  ],
  [
    "x",
    "substituting variable x",
  ],
  [
    "5",
    "created real",
  ],
]
`;

exports[`invoke of specific node types returns unbound variables directly 1`] = `
[
  [
    "(x)()",
    "identified invocation",
  ],
  [
    "([x])()",
    "processing expression",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "({x})()",
    "processed expression",
  ],
  [
    "(x)()",
    "variable invocation",
  ],
  [
    "{}",
    "established scope",
  ],
  [
    "x",
    "no substitution found in scope for variable x",
  ],
  [
    "x",
    "referenced variable",
  ],
]
`;

exports[`invoke of specific node types substitutes a variable with an unbound variable argument 1`] = `
[
  [
    "(x)(y)",
    "identified invocation",
  ],
  [
    "([x])(y)",
    "processing expression",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "({x})(y)",
    "processed expression",
  ],
  [
    "(x)([y])",
    "processing 1st",
  ],
  [
    "y",
    "referenced variable",
  ],
  [
    "(x)({y})",
    "processed 1st",
  ],
  [
    "(x)(y)",
    "variable invocation",
  ],
  [
    "{x:=y}",
    "established scope",
  ],
  [
    "x",
    "substituting variable x",
  ],
  [
    "y",
    "referenced variable",
  ],
]
`;

exports[`invoke when given an expression with multiple instances of a variable substitutes a newly bound value for each variable instance 1`] = `
[
  [
    "((x*(x+5)))(4)",
    "identified invocation",
  ],
  [
    "([(x*(x+5))])(4)",
    "processing expression",
  ],
  [
    "(x*(x+5))",
    "identified multiplication",
  ],
  [
    "([x]*(x+5))",
    "processing left operand",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "({x}*[(x+5)])",
    "processed left operand; processing right operand",
  ],
  [
    "(x+5)",
    "identified addition",
  ],
  [
    "([x]+5)",
    "processing left operand",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "({x}+[5])",
    "processed left operand; processing right operand",
  ],
  [
    "5",
    "created real",
  ],
  [
    "(x+{5})",
    "processed right operand",
  ],
  [
    "(x+5)",
    "created addition",
  ],
  [
    "(x*{(x+5)})",
    "processed right operand",
  ],
  [
    "(x*(x+5))",
    "created multiplication",
  ],
  [
    "({(x*(x+5))})(4)",
    "processed expression",
  ],
  [
    "((x*(x+5)))([4])",
    "processing 1st",
  ],
  [
    "4",
    "created real",
  ],
  [
    "((x*(x+5)))({4})",
    "processed 1st",
  ],
  [
    "((x*(x+5)))(4)",
    "multiplication invocation",
  ],
  [
    "{x:=4}",
    "established scope",
  ],
  [
    "(x*(x+5))",
    "looking for variables to substitute in expression",
  ],
  [
    "(x*(x+5))",
    "identified multiplication",
  ],
  [
    "([x]*(x+5))",
    "processing left operand",
  ],
  [
    "x",
    "substituting variable x",
  ],
  [
    "4",
    "created real",
  ],
  [
    "({4}*[(x+5)])",
    "processed left operand; processing right operand",
  ],
  [
    "(x+5)",
    "looking for variables to substitute in expression",
  ],
  [
    "(x+5)",
    "identified addition",
  ],
  [
    "([x]+5)",
    "processing left operand",
  ],
  [
    "x",
    "substituting variable x",
  ],
  [
    "4",
    "created real",
  ],
  [
    "({4}+[5])",
    "processed left operand; processing right operand",
  ],
  [
    "5",
    "invoked Real",
  ],
  [
    "5",
    "created real",
  ],
  [
    "(4+{5})",
    "processed right operand",
  ],
  [
    "(4+5)",
    "real addition",
  ],
  [
    "9",
    "created real",
  ],
  [
    "(4*{9})",
    "processed right operand",
  ],
  [
    "(4*9)",
    "real multiplication",
  ],
  [
    "36",
    "created real",
  ],
]
`;

exports[`invoke with a passed scope and arguments evaluates any unbound variables against argument overridden scope 1`] = `
[
  [
    "((x+y))(10)",
    "identified invocation",
  ],
  [
    "([(x+y)])(10)",
    "processing expression",
  ],
  [
    "(x+y)",
    "identified addition",
  ],
  [
    "([x]+y)",
    "processing left operand",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "({x}+[y])",
    "processed left operand; processing right operand",
  ],
  [
    "y",
    "referenced variable",
  ],
  [
    "(x+{y})",
    "processed right operand",
  ],
  [
    "(x+y)",
    "created addition",
  ],
  [
    "({(x+y)})(10)",
    "processed expression",
  ],
  [
    "((x+y))([10])",
    "processing 1st",
  ],
  [
    "10",
    "created real",
  ],
  [
    "((x+y))({10})",
    "processed 1st",
  ],
  [
    "((x+y))(10)",
    "addition invocation",
  ],
  [
    "{x:=10}",
    "established scope",
  ],
  [
    "(x+y)",
    "looking for variables to substitute in expression",
  ],
  [
    "(x+y)",
    "identified addition",
  ],
  [
    "([x]+y)",
    "processing left operand",
  ],
  [
    "x",
    "substituting variable x",
  ],
  [
    "10",
    "created real",
  ],
  [
    "({10}+[y])",
    "processed left operand; processing right operand",
  ],
  [
    "y",
    "no substitution found in scope for variable y",
  ],
  [
    "y",
    "referenced variable",
  ],
  [
    "(10+{y})",
    "processed right operand",
  ],
  [
    "(10+y)",
    "reorder operands",
  ],
  [
    "(y+10)",
    "identified addition",
  ],
  [
    "([y]+10)",
    "processing left operand",
  ],
  [
    "y",
    "referenced variable",
  ],
  [
    "({y}+[10])",
    "processed left operand; processing right operand",
  ],
  [
    "10",
    "created real",
  ],
  [
    "(y+{10})",
    "processed right operand",
  ],
  [
    "(y+10)",
    "created addition",
  ],
]
`;

exports[`invoke with a passed scope but no arguments evaluates any unbound variables against its scope 1`] = `
[
  [
    "((x+y))()",
    "identified invocation",
  ],
  [
    "([(x+y)])()",
    "processing expression",
  ],
  [
    "(x+y)",
    "identified addition",
  ],
  [
    "([x]+y)",
    "processing left operand",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "({x}+[y])",
    "processed left operand; processing right operand",
  ],
  [
    "y",
    "referenced variable",
  ],
  [
    "(x+{y})",
    "processed right operand",
  ],
  [
    "(x+y)",
    "created addition",
  ],
  [
    "({(x+y)})()",
    "processed expression",
  ],
  [
    "((x+y))()",
    "addition invocation",
  ],
  [
    "{x:=5}",
    "established scope",
  ],
  [
    "(x+y)",
    "looking for variables to substitute in expression",
  ],
  [
    "(x+y)",
    "identified addition",
  ],
  [
    "([x]+y)",
    "processing left operand",
  ],
  [
    "x",
    "substituting variable x",
  ],
  [
    "5",
    "created real",
  ],
  [
    "({5}+[y])",
    "processed left operand; processing right operand",
  ],
  [
    "y",
    "no substitution found in scope for variable y",
  ],
  [
    "y",
    "referenced variable",
  ],
  [
    "(5+{y})",
    "processed right operand",
  ],
  [
    "(5+y)",
    "reorder operands",
  ],
  [
    "(y+5)",
    "identified addition",
  ],
  [
    "([y]+5)",
    "processing left operand",
  ],
  [
    "y",
    "referenced variable",
  ],
  [
    "({y}+[5])",
    "processed left operand; processing right operand",
  ],
  [
    "5",
    "created real",
  ],
  [
    "(y+{5})",
    "processed right operand",
  ],
  [
    "(y+5)",
    "created addition",
  ],
]
`;

exports[`invoke with no passed scope applies its arguments in appearance order 1`] = `
[
  [
    "((x+y))(5)",
    "identified invocation",
  ],
  [
    "([(x+y)])(5)",
    "processing expression",
  ],
  [
    "(x+y)",
    "identified addition",
  ],
  [
    "([x]+y)",
    "processing left operand",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "({x}+[y])",
    "processed left operand; processing right operand",
  ],
  [
    "y",
    "referenced variable",
  ],
  [
    "(x+{y})",
    "processed right operand",
  ],
  [
    "(x+y)",
    "created addition",
  ],
  [
    "({(x+y)})(5)",
    "processed expression",
  ],
  [
    "((x+y))([5])",
    "processing 1st",
  ],
  [
    "5",
    "created real",
  ],
  [
    "((x+y))({5})",
    "processed 1st",
  ],
  [
    "((x+y))(5)",
    "addition invocation",
  ],
  [
    "{x:=5}",
    "established scope",
  ],
  [
    "(x+y)",
    "looking for variables to substitute in expression",
  ],
  [
    "(x+y)",
    "identified addition",
  ],
  [
    "([x]+y)",
    "processing left operand",
  ],
  [
    "x",
    "substituting variable x",
  ],
  [
    "5",
    "created real",
  ],
  [
    "({5}+[y])",
    "processed left operand; processing right operand",
  ],
  [
    "y",
    "no substitution found in scope for variable y",
  ],
  [
    "y",
    "referenced variable",
  ],
  [
    "(5+{y})",
    "processed right operand",
  ],
  [
    "(5+y)",
    "reorder operands",
  ],
  [
    "(y+5)",
    "identified addition",
  ],
  [
    "([y]+5)",
    "processing left operand",
  ],
  [
    "y",
    "referenced variable",
  ],
  [
    "({y}+[5])",
    "processed left operand; processing right operand",
  ],
  [
    "5",
    "created real",
  ],
  [
    "(y+{5})",
    "processed right operand",
  ],
  [
    "(y+5)",
    "created addition",
  ],
]
`;

exports[`invoke with no passed scope applies its arguments to unbound variables 1`] = `
[
  [
    "((x+y))(5,10)",
    "identified invocation",
  ],
  [
    "([(x+y)])(5,10)",
    "processing expression",
  ],
  [
    "(x+y)",
    "identified addition",
  ],
  [
    "([x]+y)",
    "processing left operand",
  ],
  [
    "x",
    "referenced variable",
  ],
  [
    "({x}+[y])",
    "processed left operand; processing right operand",
  ],
  [
    "y",
    "referenced variable",
  ],
  [
    "(x+{y})",
    "processed right operand",
  ],
  [
    "(x+y)",
    "created addition",
  ],
  [
    "({(x+y)})(5,10)",
    "processed expression",
  ],
  [
    "((x+y))([5],10)",
    "processing 1st",
  ],
  [
    "5",
    "created real",
  ],
  [
    "((x+y))({5},[10])",
    "processed 1st; processing 2nd",
  ],
  [
    "10",
    "created real",
  ],
  [
    "((x+y))(5,{10})",
    "processed 2nd",
  ],
  [
    "((x+y))(5,10)",
    "addition invocation",
  ],
  [
    "{x:=5,y:=10}",
    "established scope",
  ],
  [
    "(x+y)",
    "looking for variables to substitute in expression",
  ],
  [
    "(x+y)",
    "identified addition",
  ],
  [
    "([x]+y)",
    "processing left operand",
  ],
  [
    "x",
    "substituting variable x",
  ],
  [
    "5",
    "created real",
  ],
  [
    "({5}+[y])",
    "processed left operand; processing right operand",
  ],
  [
    "y",
    "substituting variable y",
  ],
  [
    "10",
    "created real",
  ],
  [
    "(5+{10})",
    "processed right operand",
  ],
  [
    "(5+10)",
    "real addition",
  ],
  [
    "15",
    "created real",
  ],
]
`;
